상품 목록 페이지 좌측 카테고리 및 필터 UI

HelloHailie opened the pull request:
## #️⃣ Issue Number

<!--- ex) #이슈번호, #이슈번호 -->

## 📝 요약(Summary)

- 상품 목록 페이지 전체 레이아웃 작업
좌측 옵션
- 카테고리
- 필터

## 🛠️ PR 유형

어떤 변경 사항이 있나요?

- [x] 새로운 기능 추가
- [ ] 버그 수정
- [ ] CSS 등 사용자 UI 디자인 변경
- [ ] 코드에 영향을 주지 않는 변경사항(오타 수정, 탭 사이즈 변경, 변수명 변경)
- [ ] 코드 리팩토링
- [ ] 주석 추가 및 수정
- [ ] 문서 수정
- [ ] 테스트 추가, 테스트 리팩토링
- [ ] 빌드 부분 혹은 패키지 매니저 수정
- [ ] 파일 혹은 폴더명 수정
- [ ] 파일 혹은 폴더 삭제

## 📸스크린샷 (선택)

![Jan-19-2025 14-55-14](https://github.com/user-attachments/assets/38b4f13e-85d5-4c2f-9eae-224cced5b982)

## 💬 공유사항 to 리뷰어

임의의 mock 데이터로 UI 구현하여 추후 msw 작업이 필요한 점 참고 부탁드립니다. (다른 브랜치에서 작업 예정)
고민하다가 color는 and 조건으로 들어가도록 해보았는데 의견 있으시다면 말씀 부탁드립니다. 

<!--- 리뷰어가 중점적으로 봐줬으면 좋겠는 부분이 있으면 적어주세요. -->
<!--- 논의해야할 부분이 있다면 적어주세요.-->
<!--- ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요? -->

## ✅ PR Checklist

PR이 다음 요구 사항을 충족하는지 확인하세요.

- [ ] 커밋 메시지 컨벤션에 맞게 작성했습니다.
- [ ] 변경 사항에 대한 테스트를 했습니다.(버그 수정/기능에 대한 테스트).

HelloHailie committed:
feat: CategoryList 및 Filter 컴포넌트 추가

HelloHailie committed:
feat: 상품 목록 페이지 추가

HelloHailie committed:
feat: 색상 데이터 추가 및 마크업

HelloHailie committed:
feat: 가격 range slider 추가

HelloHailie committed:
feat: Filter 컴포넌트 상태 관리 로직 개선

- Filter 컴포넌트 내부 상태 관리로 리팩토링
- 가격 범위와 색상 선택 기능의 독립적인 상태 관리 구현
- 선택된 필터 옵션을 시각적으로 표시하는 UI 추가
- 불필요한 props 제거

HelloHailie committed:
refactor: Filter 컴포넌트 구조 개선 및 컴포넌트 분리

- Filter 컴포넌트를 더 작은 단위의 컴포넌트로 분리
  - SelectedOptionTag: 선택된 필터 옵션 표시
  - PriceFilter: 가격 필터 관련 UI/로직
  - ColorFilter: 색상 필터 관련 UI/로직
- 스타일링 일관성 개선 (border 스타일 통일)

HelloHailie committed:
feat: 카테고리 ui 개선

kigpand commented on src/app/products/_components/Filter.tsx:
Filter 컴포넌트를 분리하지않고 한곳에 쓰신 이유가 있을까요?
리뷰하는 입장에서는 코드가 길어서 리뷰하기가 좀 힘든것같습니다

HelloHailie replied:
컴포넌트로 나누는 연습이 덜되서 한꺼번에 몰아서 써버렸는데요, 피드백 주셔서 filter 폴더 하위로 나눠서 작성해보았습니다. 

추가로 app/products/ 하위에 _components를 추가한 이유는 공통으로 사용할만한 컴포넌트가 아니라서 추가했는데 보시고 피드백 부탁드립니다.

HelloHailie committed:
refactor: Filter 컴포넌트 구조 개선

- Filter 컴포넌트를 독립적인 디렉토리로 이동
- Filter 컴포넌트 경로 변경에 따른 import 경로 수정
- mockProducts를 props로 전달하도록 변경

kigpand commented on src/app/products/_components/filter/Filter.tsx:
priceRange 부분 함수로 만들어서 리턴하실꺼면 map부분을 하나의 변수로 통일하면 map 중복 호출을 피할 수 있을거 같아용!

HelloHailie replied:
맵핑이 두번 되고 있어서 중복되지 않도록 수정했습니다. 감사합니다.

kigpand commented on src/hooks/useFilterOptions.ts:
useEffect 내부 코드가 복잡한 로직을 처리하는거 같은데 useMemo를 활용해보면 좋을거같아요!

HelloHailie replied:
말씀주신대로 수정했습니다.

HelloHailie committed:
fix: 이미 계산된 priceRangeValues 전달로 불필요한 map 연산 제거

HelloHailie committed:
fix: 불필요한 상태 업데이트가 제거 및 메모이제이션을 통해 성능 개선

DEV4N4 commented on src/app/products/_components/filter/Filter.tsx:
```diff
@@ -0,0 +1,71 @@
+    min: Math.min(...products.map((p) => p.price)),
+    max: Math.max(...products.map((p) => p.price)),
+  }));
+  const [selectedPriceRange, setSelectedPriceRange] = useState<PriceRange | undefined>(undefined);
```
선민님! 여기 null이 아닌 undefined를 사용하신 이유가 있으신가요?

DEV4N4 commented on package.json:
```diff
@@ -11,6 +11,7 @@
   },
   "dependencies": {
     "@hookform/resolvers": "^3.9.1",
+    "@radix-ui/react-slider": "^1.2.2",
```
엇 shadcn이 아닌 radix에서 slider를 가져오셨네요! 찾아보니 shadcn에도 slider가 있던데 radix를 사용하신 이유를 여쭤봐도 될까요?

DEV4N4 commented on src/app/products/_components/filter/Filter.tsx:
```diff
@@ -0,0 +1,71 @@
+  }));
+  const [selectedPriceRange, setSelectedPriceRange] = useState<PriceRange | undefined>(undefined);
+  const availableOptions = useFilterOptions(products);
+  const [sliderValue, setSliderValue] = useState([priceRange.min, priceRange.max]);
```
sliderValue의 state 타입을 지정하지 않으셨는데 혹시 [number, number] 외의 값이 올 것을 염두에 두신 건가요?
다른 곳에서 sliderValue나 setSliderValue가 사용되는 것을 보니 대체로 [number, number] 형식의 값만 들어오는 것 같은데 다르게 들어올 가능성이 없다면 타입을 명확히 하는게 더 좋지 않을까 하는 생각이 들었습니다!

DEV4N4 commented on src/app/products/_components/filter/Filter.tsx:
```diff
@@ -0,0 +1,71 @@
+  const [selectedColors, setSelectedColors] = useState<string[]>([]);
+
+  const handlePriceSearch = () => setSelectedPriceRange(priceRange);
+  const handleSliderChange = (value: number[]) => {
+    setSliderValue(value);
+    setPriceRange({ min: value[0], max: value[1] });
+  };
```
```suggestion
  const handleSliderChange = ([min, max]: [number, number]) => {
    setSliderValue([min, max]);
    setPriceRange({ min, max });
  };
```

혹시 이렇게 짜는건 어떠신가요? value에 [number, number] 형식의 값이 들어오는 것 같아 제안드립니다!

DEV4N4 commented on src/app/products/_components/filter/constants.ts:
```diff
@@ -0,0 +1,10 @@
+  Cream: '#F5E6D3',
+  'Space Gray': '#4A4A4A',
+  Silver: '#E3E4E5',
+};
```
이건 그냥 궁금한건데 컬러 이름은 선민님께서 정하신건가요? 😃  아니면 정해주는 플러그인이 있나요?

DEV4N4 reviewed (commented):
코드 잘 읽었습니다! 깔끔하게 짜시려고 신경쓰신 부분들이 보여서 좋았어요 😀  많이 배웠습니다!
